runtimes/common/src/bonded_currencies/mod.rs

```diff
@@ -0,0 +1,305 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2024 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+pub mod hooks;
+
+use frame_support::{
+	dispatch::DispatchResult,
+	traits::{
+		fungible::{self, UnionOf},
+		fungibles::{self, metadata::Inspect},
+		tokens::{
+			AssetId as AssetIdTraits, DepositConsequence, Fortitude, Precision, Preservation, Provenance,
+			WithdrawConsequence,
+		},
+		AccountTouch, Get,
+	},
+};
+use sp_runtime::{
+	traits::Convert,
+	DispatchError, Either,
+	Either::{Left, Right},
+};
+use sp_std::{marker::PhantomData, vec::Vec};
+use substrate_fixed::types::{I75F53, U75F53};
+use xcm::v4::Location;
+
+/// The AssetId for bonded assets.
+pub type AssetId = u32;
+
+/// Fixed point number used for creating bonding curves.
+pub type FloatInput = U75F53;
+
+/// Fixed point number used for doing calculation steps in the bonding curves.
+pub type Float = I75F53;
+
+pub struct TargetFromLeft<Target, L = Location>(PhantomData<(Target, L)>);
+impl<Target: Get<L>, L: PartialEq + Eq> Convert<L, Either<(), L>> for TargetFromLeft<Target, L> {
+	fn convert(l: L) -> Either<(), L> {
+		Target::get().eq(&l).then(|| Left(())).map_or(Right(l), |n| n)
```

Comment:
This definitely could use some comments!

Reply:
Added some explanation in: [57155ad](https://github.com/KILTprotocol/kilt-node/pull/818/commits/57155adffd2905b964e38a3d2e5b9d44e48c5dd5)